### PR TITLE
first round of changes to expose streams to the log unit

### PR DESCRIPTION
**Pull request by no2chem**
This PR introduces the first round of changes which will
expose the contents of each stream to the logging units.

The log unit interface has been simplified to take a
single parameter, which is the log data itself.

Managining serialization during a write to multiple
logging units has also been enhanced.

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/protocols/logprotocol/StreamedLogData.java [hunk 1]
```diff
@@ -1,48 +1,142 @@
...
+    @Override
+    public boolean containsStream(UUID stream) {
+        return streamDataMap.containsKey(stream);
     }
 
+    /** {@inheritDoc} */
     @Override
-    public Object getPayload(CorfuRuntime t) {
-        return null;
+    public Map<UUID, Long> getBackpointerMap() {
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/protocols/logprotocol/StreamedLogData.java]: I can't imagine why we would need this method in the future; but perhaps we need it in the interim?

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/WriteRequest.java [hunk 2]
```diff
@@ -38,6 +38,11 @@ public WriteRequest(WriteMode writeMode, DataType dataType,  Map<UUID, Long> str
         this.data = new LogData(dataType, buf);
     }
 
+    public WriteRequest(ILogData data) {
+        writeMode = WriteMode.NORMAL;
+        this.data = data;
+        streamAddresses = null;
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/protocols/wireprotocol/WriteRequest.java]: can we get rid of streamAddresses?

with quorum replication, we are not supporting Replex anyway.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **6bee0cd**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Memory mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #532 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/oqdFxm4FT5qW9s7eIIWy9kQWjSV8WgmS)

File: runtime/src/main/java/org/corfudb/protocols/logprotocol/StreamedLogData.java (hunk 1 above)
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/protocols/logprotocol/StreamedLogData.java]: We do need it in the interim.

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/WriteRequest.java (hunk 2 above)
Comment (no2chem) [inline runtime/src/main/java/org/corfudb/protocols/wireprotocol/WriteRequest.java]: Would be nice, but it's pretty entrenched in the code base. I thought the goal was small PRs. :)

File: runtime/src/main/java/org/corfudb/protocols/logprotocol/StreamedLogData.java [hunk 3]
```diff
@@ -1,48 +1,142 @@
...
     @Getter
     public final @Nonnull
-    Map<UUID, StreamData> entryMap;
+    Map<UUID, StreamData> streamDataMap;
 
-    public StreamedLogData(Map<UUID, StreamData> entryMap) {
-        this.entryMap = entryMap;
+    /** The metadata for this log entry. */
+    @Getter
+    final EnumMap<LogUnitMetadataType, Object> metadataMap;
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/protocols/logprotocol/StreamedLogData.java]: Do we not want to specify the metadata type ?

File: runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java [hunk 4]
```diff
@@ -84,7 +86,7 @@ public ILogData peek(long globalAddress) {
      * @param data          The data to propagate, or NULL,
      *                      if it is to be a hole.
      */
-    protected void propagate(long globalAddress, @Nullable ByteBuf data) {
+    protected void propagate(long globalAddress, @Nullable ILogData data) {
         int numUnits = layout.getSegmentLength(globalAddress);
 
         for (int i = 1; i < numUnits; i++) {
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java]: propagate starts from i = 1.
and recover also starts from i = 1 even if the failure could have been at the second log unit. This seems to be inefficient.

File: runtime/src/main/java/org/corfudb/runtime/clients/LogUnitClient.java [hunk 5]
```diff
@@ -234,7 +234,15 @@ private static Object handleReadDataCorruption(CorfuPayloadMsg<ReadResponse> msg
...
 
-
+    /**
+     * Asynchronously write to the logging unit.
+     * @param payload   The log data to write to the logging unit.
+     * @return          A CompletableFuture which will complete with the WriteResult once the
+     *                  write completes.
+     */
+    public CompletableFuture<Boolean> write(ILogData payload) {
+        return router.sendMessageAndGetCompletable(CorfuMsgType.WRITE.payloadMsg(new WriteRequest(payload)));
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/clients/LogUnitClient.java]: I did check the StreamLog code and it does not take into account that if an address already has hole fill data and the  new write request is for a hole fill, returning success would be a good optimization but it throws OverwriteException.
Something to consider as hole fills might have a tendency to come from many clients concurrently.

File: runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java (hunk 4 above)
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java]: good point. it should be easy to fix -- let proparte take a start-index parameter.

File: runtime/src/main/java/org/corfudb/runtime/clients/LogUnitClient.java (hunk 5 above)
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/runtime/clients/LogUnitClient.java]: yes.

this is not blocking us currently, so we can go ahead and merge this PR, but definitely keep this enhancement under consideration for the future.

**Pull request merged**